[AC-1682] Flexible collections: EF data migrations for deprecated permissions

r-tome opened the pull request:
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
With Flexible Collections, the Manager role and 'EditAssignedCollections' permission are deprecated. Therefore, we must explicitly grant access to those users for the collections they previously had access to, but now with the new Manage permission. Additionally, users and groups with the AccessAll permission will now have access to all collections within their organizations.


## Code changes

This PR only contains EF data migrations, SQL migrations are [here](https://github.com/bitwarden/server/pull/3437).

Migrations for MySql, Postgres and Sqlite follow the same structure:
* **util/*/HelperScripts/2023-12-06_00_AccessAllCollectionGroups.sql:** Add entries to [dbo].[CollectionGroup] with [Manage] = 1 for all groups with 'AccessAll' permission
* **util/*/HelperScripts/2023-12-06_01_AccessAllCollectionUsers.sql:** Add entries to [dbo].[CollectionUser] with [Manage] = 1 for all organization users with 'AccessAll' permission
* **util/*/HelperScripts/2023-12-06_02_ManagersEditAssignedCollectionUsers.sql:** Update [dbo].[CollectionUser] with [Manage] = 1 for all users with Manager role or 'EditAssignedCollections' permission or access through Groups
* **util/*/Migrations/*:** An EF migration for each migration script

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

r-tome commented:
@rkac-bw please check the SQL script files for MySQL, Postgres, and SQLite. Unlike MSSQL scripts, these scripts don't include batching logic. This is because only self-hosted instances use these database types, and they have less data.

rkac-bw reviewed (approved):
Looks good to me

joseph-flinn commented:
Would it be possible to update `ac/ac-1682/data-migrations-for-deprecated-permissions` with `main` to help reduce the amount of changes to review? Most of the changes in `.github/workflows/*` look like the ones already in `main`

r-tome commented:
> Would it be possible to update `ac/ac-1682/data-migrations-for-deprecated-permissions` with `main` to help reduce the amount of changes to review? Most of the changes in `.github/workflows/*` look like the ones already in `main`

Yes, I have just updated that branch. I have also removed the extra reviewers added by this mistake. I'm sorry everyone

r-tome commented:
@rkac-bw and @eliykat the sqlite migrations are missing, I'm still working on them, hence why I put this in draft

eliykat commented on util/SqliteMigrations/HelperScripts/2024-03-22_00_EnableOrgsCollectionEnhancements.sql:
```diff
@@ -0,0 +1,151 @@
+    WHERE "CollectionId" IN (
+        SELECT "C"."Id"
+        FROM "Collection" "C"
+        INNER JOIN "TempUsersAccessAll" "TU" ON "C"."OrganizationId" = "TU"."OrganizationId"
```
Again, this is only joining on Organization.Id when I think it should also be joining with the temp table on CollectionUser.OrganizationUserId.

Please double check for this potential issue anywhere else.

r-tome replied:
I also updated the 3 scripts here.

eliykat reviewed (changes requested):
Great work!

eliykat reviewed (commented):
Also, can you please investigate rollback further as I think that's quite important in the case of a failed migration.

r-tome commented:
> Also, can you please investigate rollback further as I think that's quite important in the case of a failed migration.

When using the Sql method inside a migration, EF Core wraps it in its own transaction by default, unless specified otherwise. I learned about this now and it explains why I had to remove the transactions from Postgres. 
Ref: https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/managing?tabs=dotnet-core-cli#arbitrary-changes-via-raw-sql

r-tome commented on util/PostgresMigrations/HelperScripts/2024-03-22_00_EnableOrgsCollectionEnhancements.psql:
```diff
@@ -0,0 +1,133 @@
+﻿-- Step 1: AccessAll migration for Groups
+    -- Create a temporary table to store the groups with AccessAll = true
+    CREATE TEMPORARY TABLE IF NOT EXISTS "TempGroupsAccessAll" AS
```
I think it makes more sense to just drop the temp tables if they exist. The reasoning is if they do exist then something must have gone wrong when applying the migration the first time. If we threw an exception then we'd get stuck on a loop trying to apply the migration.

eliykat replied:
Agreed!

r-tome commented on util/PostgresMigrations/HelperScripts/2024-03-22_00_EnableOrgsCollectionEnhancements.psql:
```diff
@@ -0,0 +1,133 @@
+            "Manage" = false
+        FROM "Collection" "C"
+        WHERE "CG"."CollectionId" = "C"."Id"
+        AND "C"."OrganizationId" IN (SELECT "OrganizationId" FROM "TempGroupsAccessAll");
```
Good spot, I've updated the 3 scripts.

r-tome commented on util/MySqlMigrations/HelperScripts/2024-03-22_00_EnableOrgsCollectionEnhancements.sql:
```diff
@@ -0,0 +1,157 @@
+    ) AS `CombinedOrgUsers` ON `OU`.`Id` = `CombinedOrgUsers`.`OrganizationUserId`
+    SET `U`.`AccountRevisionDate` = UTC_TIMESTAMP();
+
+-- Step 5: Set `FlexibleCollections` = 1 for all organizations that have not yet been migrated.
+    UPDATE `Organization` AS `O`
+    JOIN (
+        SELECT DISTINCT `TG`.`OrganizationId`
+        FROM `TempGroupsAccessAll` AS `TG`
+
+        UNION
+
+        SELECT DISTINCT `TU`.`OrganizationId`
+        FROM `TempUsersAccessAll` AS `TU`
+
+        UNION
+
+        SELECT DISTINCT `OU`.`OrganizationId`
+        FROM `TempUserManagers` AS `OU`
+    ) AS `TempOrgIds` ON `O`.`Id` = `TempOrgIds`.`OrganizationId`
+    SET `O`.`FlexibleCollections` = 1
+    WHERE `O`.`FlexibleCollections` = 0;
```
You're right! I was thinking about Orgs that would be created whilst this migration started but that is such a specific scenario that is extremely unlikely to happen on a self hosted instance. I'll remove this.

justindbaur commented on util/MySqlMigrations/HelperScripts/2024-03-22_00_EnableOrgsCollectionEnhancements.sql:
```diff
@@ -33,113 +33,107 @@
             JSON_VALID(`OU`.`Permissions`) AND JSON_VALUE(`OU`.`Permissions`, '$.editAssignedCollections') = 'true'));
 
--- Start transaction
-START TRANSACTION;
```
Yep, this is my understanding too, you can check the SQL it will generate [with the CLI](https://learn.microsoft.com/en-us/ef/core/cli/dotnet#dotnet-ef-migrations-script) and see that it will add transactions. Also, when you don't add them yourself and let EF do it then it will respect the `--no-transactions` option on that command. Why you'd want to do that I have no idea but it's there 🤷.

eliykat replied:
Thanks both of you for the clarification.

rkac-bw commented on util/MySqlMigrations/HelperScripts/2024-03-22_00_EnableOrgsCollectionEnhancements.sql:
```diff
@@ -1,6 +1,7 @@
     -- Create a temporary table to store the groups with AccessAll = 1
-    CREATE TEMPORARY TABLE IF NOT EXISTS `TempGroupsAccessAll` AS
+    DROP TEMPORARY TABLE IF EXISTS `TempGroupsAccessAll`;
+    CREATE TEMPORARY TABLE `TempGroupsAccessAll` AS
```
@eliykat Temp tables are dropped after connection is dropped there is no specific need to drop the table

eliykat replied:
@r-tome maybe we don't worry about dropping the tables first then? Your call.

eliykat reviewed (commented):
lgtm other than the temp tables comment thread above, which is minor.
